pkg/component/worker/kubelet.go

```diff
@@ -248,6 +254,37 @@ func (k *Kubelet) Reconcile() error {
 // Health-check interface
 func (k *Kubelet) Healthy() error { return nil }
 
+func (k *Kubelet) prepareLocalKubeletConfig(kubeletconfig string) (string, error) {
+	var kubeletConfiguration kubeletv1beta1.KubeletConfiguration
+	err := yaml.Unmarshal([]byte(kubeletconfig), &kubeletConfiguration)
+	if err != nil {
+		return "", fmt.Errorf("can't unmarshal kubelet config: %v", err)
+	}
+
+	kubeletConfiguration.Authentication.X509.ClientCAFile = filepath.Join(k.K0sVars.CertRootDir, "ca.crt")
+	kubeletConfiguration.VolumePluginDir = k.K0sVars.KubeletVolumePluginDir
+	kubeletConfiguration.KubeReservedCgroup = "system.slice"
+	kubeletConfiguration.KubeletCgroups = "/system.slice/containerd.service"
+
+	if len(k.Taints) > 0 {
+		var taints []corev1.Taint
+		for _, taint := range k.Taints {
+			parsedTaint, err := parseTaint(taint)
+			if err != nil {
+				return "", fmt.Errorf("can't parse taints for profile config map: %v", err)
+			}
+			taints = append(taints, parsedTaint)
+		}
+		kubeletConfiguration.RegisterWithTaints = taints
+	}
+
+	localKubeletConfig, err := kubeletConfiguration.Marshal()
```

Comment:
That being said, I wonder if we should still be afraid of importing k8s.io/kubernetes. After taking a look at the upcoming go.mod changes for k8s 1.24, everything is so intertwined that it probably wouldn't make things worse than they already are.

Reply:
Yeah, but in my case, I unmarshal the yaml into a struct, then add some values, so all default fields should be in place.

Reply:
That's what I thought in the beginning, too. But unfortunately, that's not the case. [`SetObjectDefaults_KubeletConfiguration()`](https://pkg.go.dev/k8s.io/kubernetes/pkg/kubelet/apis/config/v1beta1#SetObjectDefaults_KubeletConfiguration) is required for that.

I did the following test. Given some default worker profile like this:

<details><summary>profile.yaml</summary>

```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
authentication:
  anonymous:
    enabled: false
  webhook:
    cacheTTL: 0s
    enabled: true
authorization:
  mode: Webhook
  webhook:
    cacheAuthorizedTTL: 0s
    cacheUnauthorizedTTL: 0s
cgroupsPerQOS: true
clusterDNS:
- 10.96.0.10
clusterDomain: cluster.local
eventRecordQPS: 0
failSwapOn: false
kind: KubeletConfiguration
rotateCertificates: true
serverTLSBootstrap: true
tlsCipherSuites:
- TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
- TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
- TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
- TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
- TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305
- TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
- TLS_RSA_WITH_AES_256_GCM_SHA384
- TLS_RSA_WITH_AES_128_GCM_SHA256
volumeStatsAggPeriod: 0s
```

</details>

I created a patched KubeletConfiguration like this:

```go
underTest := &Kubelet{Taints: []string{"foo=abc:NoSchedule"}}
patchedYaml, err := underTest.prepareLocalKubeletConfig(
	"<profile.yaml>",
	kubeletConfig{
		ClientCAFile:       "ClientCAFile",
		VolumePluginDir:    "VolumePluginDir",
		KubeReservedCgroup: "KubeReservedCgroup",
		KubeletCgroups:     "KubeletCgroups",
		CgroupsPerQOS:      true,
		ResolvConf:         "ResolvConf",
	},
)
```

This resulted in the following YAML:

<details><summary>patched.yaml</summary>

```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
authentication:
  anonymous:
    enabled: false
  webhook:
    cacheTTL: 0s
    enabled: true
  x509:
    clientCAFile: ClientCAFile
authorization:
  mode: Webhook
  webhook:
    cacheAuthorizedTTL: 0s
    cacheUnauthorizedTTL: 0s
cgroupsPerQOS: true
clusterDNS:
- 10.96.0.10
clusterDomain: cluster.local
cpuManagerReconcilePeriod: 0s
eventRecordQPS: 0
evictionPressureTransitionPeriod: 0s
failSwapOn: false
fileCheckFrequency: 0s
httpCheckFrequency: 0s
imageMinimumGCAge: 0s
kind: KubeletConfiguration
kubeReservedCgroup: KubeReservedCgroup
kubeletCgroups: KubeletCgroups
logging:
  flushFrequency: 0
  options:
    json:
      infoBufferSize: "0"
  verbosity: 0
memorySwap: {}
nodeStatusReportFrequency: 0s
nodeStatusUpdateFrequency: 0s
registerWithTaints:
- effect: NoSchedule
  key: foo
  value: abc
resolvConf: ResolvConf
rotateCertificates: true
runtimeRequestTimeout: 0s
serverTLSBootstrap: true
shutdownGracePeriod: 0s
shutdownGracePeriodCriticalPods: 0s
streamingConnectionIdleTimeout: 0s
syncFrequency: 0s
tlsCipherSuites:
- TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
- TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
- TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
- TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
- TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305
- TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
- TLS_RSA_WITH_AES_256_GCM_SHA384
- TLS_RSA_WITH_AES_128_GCM_SHA256
volumePluginDir: VolumePluginDir
volumeStatsAggPeriod: 0s
```

</details>

The diff reveals the problems:

<details><summary>diff profile.yaml patched.yaml</summary>

```diff
--- profile.yaml
+++ patched.yaml
@@ -5,6 +5,8 @@
   webhook:
     cacheTTL: 0s
     enabled: true
+  x509:
+    clientCAFile: ClientCAFile
 authorization:
   mode: Webhook
   webhook:
@@ -14,11 +16,37 @@
 clusterDNS:
 - 10.96.0.10
 clusterDomain: cluster.local
+cpuManagerReconcilePeriod: 0s
 eventRecordQPS: 0
+evictionPressureTransitionPeriod: 0s
 failSwapOn: false
+fileCheckFrequency: 0s
+httpCheckFrequency: 0s
+imageMinimumGCAge: 0s
 kind: KubeletConfiguration
+kubeReservedCgroup: KubeReservedCgroup
+kubeletCgroups: KubeletCgroups
+logging:
+  flushFrequency: 0
+  options:
+    json:
+      infoBufferSize: "0"
+  verbosity: 0
+memorySwap: {}
+nodeStatusReportFrequency: 0s
+nodeStatusUpdateFrequency: 0s
+registerWithTaints:
+- effect: NoSchedule
+  key: foo
+  value: abc
+resolvConf: ResolvConf
 rotateCertificates: true
+runtimeRequestTimeout: 0s
 serverTLSBootstrap: true
+shutdownGracePeriod: 0s
+shutdownGracePeriodCriticalPods: 0s
+streamingConnectionIdleTimeout: 0s
+syncFrequency: 0s
 tlsCipherSuites:
 - TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
 - TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
@@ -28,4 +56,5 @@
 - TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
 - TLS_RSA_WITH_AES_256_GCM_SHA384
 - TLS_RSA_WITH_AES_128_GCM_SHA256
+volumePluginDir: VolumePluginDir
 volumeStatsAggPeriod: 0s
```

</details>

The following fields have a default value that's not the empty value:

* [`cpuManagerReconcilePeriod 10s`](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L351)
* [`evictionPressureTransitionPeriod 5m`](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L500)
* [`fileCheckFrequency 20s`](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L107)
* [`httpCheckFrequency 20s`](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L111)
* [`imageMinimumGCAge 2m`](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L288)
* [`nodeStatusReportFrequency 5m`](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L273)
* [`nodeStatusUpdateFrequency 10s`](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L263)
* [`runtimeRequestTimeout 2m`](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L394)
* [`streamingConnectionIdleTimeout 4h`](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L255)
* [`syncFrequency 1m`](https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L102)

Reply:
But it is exactly what kubelet does during the config parsing. It fills the given config with the default values.

Reply:
Today I learned: Instead of parsing YAML data into a defaulted struct, Kubernetes [parses the data into an empty struct][p] and [applies its defaulters afterwards][d].

[p]: https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L139
[d]: https://github.com/kubernetes/kubernetes/blob/v1.24.0/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L187

That means that for fields where default value  != zero value, in YAML terms, both the absence of the value as well as the zero value will result in the default value being used. Unless the value is represented as a pointer in the struct, in which case the zero value is used instead of the default value, but you need to know this because there is nothing in the YAML that will tell you that. That's such a non-intuitive behavioral quirk.

As an example, no fileCheckFrequency at all, `fileCheckFrequency: 0s` and `fileCheckFrequency: 20s` are all semantically equivalent and will result in the Kubelet using a value 20 seconds. Wow.

Technically everything is fine here, but there is a chance that some users will be as confused as I was when they look at the contents of the resulting Kubelet configuration file. Especially when troubleshooting a cluster, all those zero values look very suspicious to the uninformed. You have to go all the way down into source code, to the definition of the struct that backs the CRD to actually figure out what value will be in effect.

Reply:
Yeah, that's pretty unclear. But at least with this change, we will have on the filesystem the same config that is really being used in the kubelet.